File: Array/10. MissingNumberInArray.ts
function missingNumber(nums: number[]): number {
  let n = nums.length;

  for (let i = 0; i <= n; i++) {
    let flag = 0;
    for (let j = 0; j < n; j++) {
      if (nums[j] === i) {
        flag = 1;
        break;
      }
    }
    if (flag === 0) {
      return i;
    }
  }

  return -1;
}

function missingNumber2(nums: number[]): number {
  let n = nums.length;
  let hashedArray = new Array(n + 1).fill(-1);

  for (let i = 0; i < n; i++) {
    hashedArray[nums[i]]++;
  }

  for (let i = 0; i < hashedArray.length; i++) {
    if (hashedArray[i] === -1) {
      return i;
    }
  }
  return -1;
}

function missingNumber3(nums: number[]): number {
  let n = nums.length;

  let s1 = (n * (n + 1)) / 2;
  let s2 = 0;

  for (let i = 0; i < n; i++) {
    s2 += nums[i];
  }

  return s1 - s2;
}

// console.log(missingNumber2([0]));
// console.log(missingNumber3([3, 0, 1]));
// console.log(missingNumber([0, 1, 2, 4, 5]));
// console.log(missingNumber([0, 1]));
// console.log(missingNumber3([9, 6, 4, 2, 3, 5, 7, 0, 1]));
console.log(missingNumber3([0, 1, 2, 3, 4]));
// console.log(missingNumber([1]));
// console.log(missingNumber([1, 2]));

File: Array/10. MissingNumberInArray.py
def brute_force_approach(nums):
    n = len(nums)+1
    for i in range(1, n+1):
        flag = 0
        for j in range(0, n-1):
            if nums[j] == i:
                flag = 1
                break

        if flag == 0:
            return i

    return -1


# print(brute_force_approach([1, 2, 4, 5, 6]))
# print(brute_force_approach([3, 0, 1]))
# print(brute_force_approach([0, 1]))
# print(brute_force_approach([9, 6, 4, 2, 3, 5, 7, 0, 1]))

def better_approach(nums):
    n = len(nums)
    hashed_array = [0]*(n+1)
    for i in range(0, n):
        hashed_array[nums[i]] += 1

    for j in range(1, len(hashed_array)):
        if hashed_array[j] == 0:
            return j

    return -1


# print(better_approach([9, 6, 4, 2, 3, 5, 7, 0, 1]))
# print(better_approach([3, 0, 1]))

def optimized_approach(nums):
    lo, hi = min(nums), max(nums)

    s1 = (lo + hi) * (hi - lo + 1) // 2 
    s2 =  sum(nums)
    
    return s1 - s2

print(optimized_approach([8,2,4,5,3,7,1]))
print(optimized_approach([1, 2, 3, 5]))
print(optimized_approach([9, 6, 4, 2, 3, 5, 7, 0, 1]))
print(optimized_approach([3, 0, 1]))